### PR TITLE
Adding 3D columnar option in PolycrystalHex userobject

### PR DESCRIPTION
**Pull request by sourabhkadambi**
Adding option to generate 3D columnar hexagonal grains with centroids positioned at the middle of columnar dimension.
Addresses #25159

### CODE REVIEW

**Comment by sourabhkadambi** [comment]
@GiudGiud

File: modules/phase_field/test/tests/grain_growth/hex_columnar_3D_out.e
Comment (GiudGiud) [inline modules/phase_field/test/tests/grain_growth/hex_columnar_3D_out.e:1]: we ll want to check only the intiial condition, and on a small mesh as well

**Comment by GiudGiud** [comment]
@laagesen @permcody or @hugary1995 anyone to review this?
I can do the editing of the submission if either of you will review it after (and we get a final ccb stamp after)

**Comment by hugary1995** [comment]
Error out if dim == 3 and columnar_3D == false. Otherwise looks good to. And I second @GiudGiud 's suggestion.

File: modules/phase_field/test/tests/grain_growth/hex_columnar_3D_out.e
Comment (sourabhkadambi) [reply modules/phase_field/test/tests/grain_growth/hex_columnar_3D_out.e:1]: Okay, I replaced it with an initial condition test

**Comment by GiudGiud** [comment]
Added the tests configuration file

**Comment by GiudGiud** [comment]
@permcody for the review

**Review by laagesen — CHANGES_REQUESTED** [review]
@sourabhkadambi can you please also update the documentation file `PolyCrystalHex.md` to describe the new option?

**Review by permcody — DISMISSED** [review]
Larry's most recent comment on updating the documentation is good. Otherwise, I approve. Thanks for the contribution!

**Comment by sourabhkadambi** [comment]
While updating the docs, I found issues with a couple of other things with this object. Hexes are properly generated only for even square number of grains and periodic BCs. I decided to add error messages when they are not satisfied. With these changes, I had to fix a couple of tests in phase_field that were failing. But it looks like a test in Marmot is failing as well. I think this test can also be fixed by just adding periodic BCs to it. Any suggestion on these new changes and how I should proceed? @laagesen

**Comment by laagesen** [comment]
I see the failing test is `recrystallization.DislocEnergyKernel` and is failing because the new version of the code checks to see if periodic BCs are being used, but they are not in this test. I think this was added by @SudiptaBiswas - can you check with her to see if changing that test to periodic BCs would cause any issues? If not, can you change the Marmot test to use periodic BCs? Thanks!

File: modules/phase_field/src/userobjects/PolycrystalHex.C [hunk 1]
```diff
@@ -33,21 +33,41 @@ PolycrystalHex::PolycrystalHex(const InputParameters & parameters)
...
+
+  if (_columnar_3D && _dim == 3)
+    d -= 1;
+
+  // check if periodic boundary condition is set
+  for (unsigned int j = 0; j < d; ++j)
+  {
+    for (unsigned int i = 0; i < _vars.size(); ++i)
+      if (!_mesh.isTranslatedPeriodic(_vars[i]->number(), j))
+        mooseError("PolycrystalHex works only with periodic BCs");
```
Comment (SudiptaBiswas) [inline modules/phase_field/src/userobjects/PolycrystalHex.C:58]: Why are we making it mandatory to use PBC?

**Comment by SudiptaBiswas** [comment]
> I see the failing test is `recrystallization.DislocEnergyKernel` and is failing because the new version of the code checks to see if periodic BCs are being used, but they are not in this test. I think this was added by @SudiptaBiswas - can you check with her to see if changing that test to periodic BCs would cause any issues? If not, can you change the Marmot test to use periodic BCs? Thanks!

That should be fine.

File: modules/phase_field/src/userobjects/PolycrystalHex.C (hunk 1 above)
Comment (sourabhkadambi) [reply modules/phase_field/src/userobjects/PolycrystalHex.C:58]: Phase-field models tend to use hexagonal grains to have a structure that's stable against grain growth. Without PBC, some of the grains generated are not six sided. This is likely to result in grain growth and may not serve the main purpose. Making it mandatory to use PBC can prevent users from inadvertently creating undesired microstructures

**Comment by sourabhkadambi** [comment]
> I see the failing test is `recrystallization.DislocEnergyKernel` and is failing because the new version of the code checks to see if periodic BCs are being used, but they are not in this test. I think this was added by @SudiptaBiswas - can you check with her to see if changing that test to periodic BCs would cause any issues? If not, can you change the Marmot test to use periodic BCs? Thanks!

Internal tests are now passing after changes to Marmot test

**Comment by sourabhkadambi** [comment]
> @sourabhkadambi can you please also update the documentation file `PolyCrystalHex.md` to describe the new option?

Updated documentation

File: modules/phase_field/src/userobjects/PolycrystalHex.C (hunk 1 above)
Comment (GiudGiud) [reply modules/phase_field/src/userobjects/PolycrystalHex.C:58]: @SudiptaBiswas are we good to go here?

**Comment by GiudGiud** [comment]
@laagesen @SudiptaBiswas we good to go on this?

**Comment by laagesen** [comment]
Good on my end

**Review by GiudGiud — APPROVED** [review]
Failures are unrelated. Thank you for the patch!

**Pull request merged**